--- /dev/null
+++ b/hfi1/hfi1_netdev.h
@@ -0,0 +1,64 @@
+/* SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause) */
+/*
+ * Copyright(c) 2018 Intel Corporation.
+ *
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * GPL LICENSE SUMMARY
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * BSD LICENSE
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *  - Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  - Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Intel Corporation nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+/*
+ * This file specifies an interface into hfi1's IPOIB functionality. This
+ * is a workaround for kernels that lack ib_device.alloc_rdma_netdev.
+ */
+
+#ifndef HFI1_HFI1_NETDEV_H
+#define HFI1_HFI1_NETDEV_H
+
+struct net_device *hfi1_alloc_rn(struct ib_device *device,
+				 u8 port_num,
+				 enum rdma_netdev_t type,
+				 const char *name,
+				 unsigned char name_assign_type,
+				 void (*setup)(struct net_device *));
+
+#endif /* HFI1_HFI1_NETDEV_H */
--- a/hfi1/verbs.c
+++ b/hfi1/verbs.c
@@ -69,7 +69,9 @@
 #include "fault.h"
 #include "affinity.h"
 #include "ipoib.h"
-
+#ifdef AIP
+#include "hfi1_netdev.h"
+#endif
 extern unsigned int selinux_mode;
 
 unsigned int hfi1_lkey_table_size = 16;
@@ -2079,7 +2081,7 @@
 #endif
 
 #ifdef AIP
-static struct net_device *hfi1_alloc_rn(struct ib_device *device,
+struct net_device *hfi1_alloc_rn(struct ib_device *device,
 					u8 port_num,
 					enum rdma_netdev_t type,
 					const char *name,
@@ -2117,6 +2119,9 @@
 
 	return dev;
 }
+#if defined(IFS_DEB9)
+EXPORT_SYMBOL(hfi1_alloc_rn);
+#endif
 #endif
 
 /**
@@ -2174,7 +2179,7 @@
 #endif
 
 #if defined(AIP) && !defined(IFS_DEB9)
-	/* Debian 9 kernel 4.9.0 lacks alloc_rdma_netdev */
+	/* Debian 9 kernel 4.9.0 lacks alloc_rdma_netdev, so hfi1 exports hfi_alloc_rn. */
 	ibdev->alloc_rdma_netdev = hfi1_alloc_rn; /* May not be in all distro kernels might need a guard */
 #endif
 	/* keep process mad in the driver */
